feat: add C ndarray interface and refactor implementation for `stats/base/dstdevyc`

aayush0325 opened the pull request:
## Description

> What is the purpose of this pull request?

This pull request:

-   adds C ndarray interface for `stats/base/dstdevyc`
-   refactors JavaScript implementation
-   updates tests, docs, examples and benchmarks

## Related Issues

> Does this pull request have any related issues?

No.

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

aayush0325 committed:
feat: add C ndarray interface and refactor implementation

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

aayush0325 committed:
feat: finishing up

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

aayush0325 committed:
chore: add missing decimals

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

aayush0325 commented:
good evening @kgryte, kindly go over [this commit](https://github.com/stdlib-js/stdlib/pull/4628/commits/90d576ed8a67e043344725157d9204bc72000187) to see where the missing decimals can be added as discussed in a previously merged PR, if this is something worth rectifying i can drop a PR for that

kgryte commented on lib/node_modules/@stdlib/stats/base/dstdevyc/docs/repl.txt:
```diff
@@ -46,22 +46,19 @@
     > {{alias}}( x.length, 1, x, 1 )
```
```suggestion
    > {{alias}}( x.length, 1.0, x, 1 )
```

kgryte commented on lib/node_modules/@stdlib/stats/base/dstdevyc/docs/repl.txt:
```diff
@@ -46,22 +46,19 @@
-    > var N = {{alias:@stdlib/math/base/special/floor}}( x.length / 2 );
-    > var stride = 2;
-    > {{alias}}( N, 1, x, stride )
+    > {{alias}}( 3, 1, x, 2 )
```
```suggestion
    > {{alias}}( 3, 1.0, x, 2 )
```

kgryte commented on lib/node_modules/@stdlib/stats/base/dstdevyc/docs/repl.txt:
```diff
@@ -46,22 +46,19 @@
-    > N = {{alias:@stdlib/math/base/special/floor}}( x0.length / 2 );
-    > stride = 2;
-    > {{alias}}( N, 1, x1, stride )
+    > {{alias}}( 3, 1, x1, 2 )
```
```suggestion
    > {{alias}}( 3, 1.0, x1, 2 )
```

kgryte commented on lib/node_modules/@stdlib/stats/base/dstdevyc/docs/repl.txt:
```diff
@@ -110,8 +107,7 @@
     > var x = new {{alias:@stdlib/array/float64}}( [ 1.0, -2.0, 3.0, 2.0, 5.0, -1.0 ] );
-    > var N = {{alias:@stdlib/math/base/special/floor}}( x.length / 2 );
-    > {{alias}}.ndarray( N, 1, x, 2, 1 )
+    > {{alias}}.ndarray( 3, 1, x, 2, 1 )
```
```suggestion
    > {{alias}}.ndarray( 3, 1.0, x, 2, 1 )
```

I should briefly state why we do this: namely, we use decimals to convey when an argument is expected to be integer-valued and when a value can have decimals. Here, the correction parameter need not be integer-valued. Hence, we add a decimal point. This is our way of using documentation to convey "type" information, even when everything in JS is a "double" from the Standards POV.

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
docs: add decimal

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/stats/base/dstdevyc/docs/types/index.d.ts:
```diff
@@ -59,7 +59,7 @@ interface Routine {
 	* var v = dstdevyc.ndarray( x.length, 1, x, 1, 0 );
```
Sigh. The lack of decimal points for the `correction` parameter is everywhere in this PR. This was my original mistake.

aayush0325 commented:
i'd have to go over the package in detail to see where else we can make this change, is that something that you'd want me to do?

kgryte commented:
> i'd have to go over the package in detail to see where else we can make this change, is that something that you'd want me to do?

Yeah, sorry. 😞 Now that we've gone down this route, we should be consistent, at least within this package.

aayush0325 committed:
chore: add missing decimals

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

aayush0325 committed:
chore: add missing decimals

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

aayush0325 commented:
I believe we're consistent in using decimals to show that the `correction` parameter is a double **in this package** now @kgryte, sorry if i missed anything

kgryte commented:
@aayush0325 Thank you for making the updates!

kgryte reviewed (approved):
LGTM
